Pod/Classes/VideoComponent/ARVideoManager.m

```diff
@@ -0,0 +1,151 @@
+#import <AVFoundation/AVFoundation.h>
+#import "ARVideoManager.h"
+
+
+@interface ARVideo : UIView
+@property (nonatomic, strong, readwrite) AVPlayer *player;
+@property (nonatomic, strong, readwrite) AVPlayerLayer *playerLayer;
+@end
+
+
+@implementation ARVideo
+
+- (void)setFrame:(CGRect)frame
+{
+    [super setFrame:frame];
+    self.playerLayer.frame = self.bounds;
+}
+
+// Loop
+
+- (void)playerDidReachEnd:(NSNotification *)notification
+{
+    AVPlayerItem *playerItem = [notification object];
+    [playerItem seekToTime:kCMTimeZero];
+    [self.player play];
+}
+
+// We want to know how large the video is to be able to set the player positioning ourselves
+// because we want to have a left-aligned video, instead of the default center aligned fill.
+
+- (void)observeValueForKeyPath:(NSString *)path ofObject:(id)object change:(NSDictionary *)change context:(void *) context
+{
+    // Only do the positioning logic for aspect fills
+    if (self.playerLayer.videoGravity == AVLayerVideoGravityResizeAspectFill && [path isEqualToString:@"readyForDisplay"]) {
+        CGRect videoRect = self.playerLayer.videoRect;
+        
+        if (!CGRectIsEmpty(videoRect)) {
+            // Ok, now we've got enough info to handle the sizing
+            [self updatePlayerLayerFrame];
+            
+            // Animate in the video
+            self.layer.opacity = 1;
+        }
+    } else {
+        [super observeValueForKeyPath:path ofObject:object change:change context:context];
+    }
+}
+
+- (void)updatePlayerLayerFrame
+{
+    if (!self.player.currentItem) {
+        return;
+    }
+    // Get the right size for the video
+    AVPlayerItem *item = self.player.currentItem;
+    AVPlayerItemTrack *track = [item tracks][0];
+    CGSize naturalVideoSize = track.assetTrack.naturalSize;
+    
+    // Figure our how it would fit un-cropped in aspect fill
+    CGFloat aspectRatio = naturalVideoSize.width / naturalVideoSize.height;
+    CGRect playerBounds = CGRectMake(0, 0, self.superview.bounds.size.height * aspectRatio, self.superview.bounds.size.height);
+    
+    // Don't animate setting the positioning of the video
+    [CATransaction begin];
+    [CATransaction setAnimationDuration:0];
+    self.playerLayer.frame = playerBounds;
+    [CATransaction commit];
+}
+
+// We need to do some memory management to make sure we don't keep a connection to the AVPlayer
+
+- (void)dealloc
+{
+    [self.playerLayer removeObserver:self forKeyPath:@"readyForDisplay"];
```

Comment:
Don’t refer to `self` in init* and dealloc, i.e. refer to `_playerLayer` instead.

Reply:
clever